cmd/controller/main.go

```diff
@@ -261,49 +305,82 @@ func startControllerViaLeaderElectionOrDie(runOptions controllerRunOptions) {
 			runOptions.name,
 		)
 
-		if err = machineController.Run(workerCount, stopCh); err != nil {
-			glog.Fatalf("error running controller: %v", err)
+		err := machineController.Run(workerCount, newCtx.Done())
+		if err != nil {
+			glog.Errorf("error running controller: %v", err)
+			return
 		}
+		glog.Info("machine controller has been successfully stopped")
 	}
 
-	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+	go leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
 		Lock:          &rl,
 		LeaseDuration: defaultLeaderElectionLeaseDuration,
 		RenewDeadline: defaultLeaderElectionRenewDeadline,
 		RetryPeriod:   defaultLeaderElectionRetryPeriod,
 		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
+			OnStartedLeading: runController,
 			OnStoppedLeading: func() {
-				glog.Fatalf("leaderelection lost, closing the app")
+				newCtxDone()
 			},
 		},
 	})
+
+	var g run.Group
+	{
+		g.Add(func() error {
```

Comment:
This group func looks pretty weird. I think we can find a better way to handle this. 😊

Reply:
yeah, I'm open to suggestion. One thing worth noting is the fact that there is no way to tell the leader election library to stop.